# Importing libraries 

In [18]:
# Importing libraries
import pandas as pd 
import numpy as np
import os 

path = r'C:\Users\ANITA BOADU\Achievement 4'

#Import Orders_products_combined as 
df_ords_prods_combs = pd.read_pickle(os.path.join(path, 'Data','Prepared Data', 'orders_products_combined.pkl'))

#ImportCleaned Products Dataset 
df_prods_checked = pd.read_csv(os.path.join(path, 'Data','Prepared Data', 'products_checked.csv'))

# Export data to pkl
df_ords_prods_merge.to_pickle(os.path.join(path, 'Data','Prepared Data', 'ords_prods_merge.pkl'))

In [5]:
df_ords_prods_combs.head()

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0           1   2398795        1    prior             2          3   
1           1   2398795        1    prior             2          3   
2           1   2398795        1    prior             2          3   
3           1   2398795        1    prior             2          3   
4           1   2398795        1    prior             2          3   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  7                    15.0         196                  1   
1                  7                    15.0       10258                  2   
2                  7                    15.0       12427                  3   
3                  7                    15.0       13176                  4   
4                  7                    15.0       26088                  5   

   reordered _merge  
0          1   both  
1          0   both  
2          1   both  
3          0   both  
4          1   both

######## `df_ords_prods_combs.head()` displays the first five rows of the merged orders-products DataFrame, giving a quick preview of its columns and data.  
This helps verify that the merge loaded correctly and shows what the dataset looks like.

In [6]:
df_ords_prods_combs.shape

(30356421, 12)

#######The shape `(30,356,421, 12)` means that `df_ords_prods_combs` has **30,356,421 rows** and **12 columns**.

In [8]:
df_prods_checked.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

######## df_prods_checked.head()` shows the first five rows of the cleaned products DataFrame, allowing you to quickly preview its columns and data.

In [13]:
df_prods_checked.shape

(49672, 5)

########The shape `(49,672, 5)` means that `df_prods_checked` has **49,672 rows** and **5 columns**.

########### The common merge key is product i_d

In [14]:
 df_ords_prods_merge= df_ords_prods_combs.merge(df_prods_checked, on='product_id', how='left')

######## This line merges `df_ords_prods_combs` with `df_prods_checked` using the `product_id` column.  
Using `how='left'` keeps all rows from `df_ords_prods_combs` and adds product details from `df_prods_checked` where available.

In [15]:
df_ords_prods_merge.head()

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0           1   2398795        1    prior             2          3   
1           1   2398795        1    prior             2          3   
2           1   2398795        1    prior             2          3   
3           1   2398795        1    prior             2          3   
4           1   2398795        1    prior             2          3   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  7                    15.0         196                  1   
1                  7                    15.0       10258                  2   
2                  7                    15.0       12427                  3   
3                  7                    15.0       13176                  4   
4                  7                    15.0       26088                  5   

   reordered _merge                product_name  aisle_id  department_id  \
0          1   both                        Soda      77.0            7.0   
1          0   both                  Pistachios     117.0           19.0   
2          1   both         Original Beef Jerky      23.0           19.0   
3          0   both      Bag of Organic Bananas      24.0            4.0   
4          1   both  Aged White Cheddar Popcorn      23.0           19.0   

   prices  
0     9.0  
1     3.0  
2     4.4  
3    10.3  
4     4.7

######## `df_ords_prods_merge.head()` displays the first five rows of the merged DataFrame, showing the combined orders, products, and their details.

In [16]:
df_ords_prods_merge['_merge'].value_counts()

_merge
both          30356957
left_only            0
right_only           0
Name: count, dtype: int64

####### The output shows that all rows (`30,356,957`) matched between `df_ords_prods_combs` and `df_prods_checked`, and there are **no unmatched rows** from either DataFrame.  
This means the left join successfully added product details for every order.

In [17]:
df_ords_prods_merge.shape

(30356957, 16)

#######The shape `(30,356,957, 16)` means that `df_ords_prods_merge` has **30,356,957 rows** and **16 columns** after merging orders with the cleaned product data.